assembly/runtime/collections/avlTree.ts

```diff
@@ -0,0 +1,652 @@
+            this.rootId = this.doRemove(key);
+            this.val.delete(key);
+        }
+    }
+    // alias to match rust sdk
+    remove(key: K): void {
+        this.delete(key);
+    }
+  
+  
+    /**
+     * Get a range of values from a start key (inclusive) to an end key (exclusive).
+     * If end is greater than max key, include start to max inclusive.
+     * 
+     * @param start Key for lower bound (inclusive).
+     * @param end Key for upper bound (exclusive).
+     * @returns Range of values corresponding to keys within start and end bounds.
+     */
+    values(start: K, end: K): V[] {
+        const keys = this.keys(start, end);
+        const values: V[] = [];
+        for (let i = 0; i < keys.length; ++i) {
+            const key = keys[i];
+            values.push(this.val.getSome(key));
+        }
+        return values;
+    }
+
+    /**
+     * Get a range of keys from a start key (inclusive) to an end key (exclusive).
+     * If end is greater than max key, include start to max inclusive.
+     * 
+     * @param start Key for lower bound (inclusive).
+     * @param end Key for upper bound (exclusive).
+     * @returns Range of keys within start and end bounds.
+     */
+    keys(start: K, end: K): K[] {
+        const rootNode = this.rootNode;
+        const sorted: K[] = [];
+        if (rootNode) {
+            const visited: AVLTreeNode<K>[] = [];
+
+            this.traverseLeft(start, rootNode, visited);
+
+            while (visited.length) {
+                const node = visited.pop();
+
+                if (
+                    // key must always be gte start bound
+                    node.key >= start && (
+                        // if start and end bound are equal, 
+                        // end bound becomes lte instead of strictly less than
+                        start === end ? 
+                            node.key <= end : 
+                            node.key < end
+                    )
+                ) {
+                    sorted.push(node.key);
+                }
+
+                if (node.key < end && node.right) {
+                    this.traverseLeft(start, this.node(node.right)!, visited);
+                }
+            }
+        }
+        
+        return sorted;
+    }
+
+    // TODO make private
+    traverseLeft(start: K, node: AVLTreeNode<K>, visited: AVLTreeNode<K>[]): void {
+        while (node.key >= start && node.left) {
+            visited.push(node);
+            node = this.node(node.left)!;
+        }
+        visited.push(node);
+    }
+
+    /**
+     * Get a range of entries from a start key (inclusive) to an end key (exclusive).
+     * If end is greater than max key, include start to max inclusive.
+     * 
+     * @param start Key for lower bound (inclusive).
+     * @param end Key for upper bound (exclusive).
+     * @returns Range of entries corresponding to keys within start and end bounds.
+     */
+    entries(start: K, end: K): MapEntry<K, V>[] {
```

Comment:
```suggestion
    entries(start: K = this.min(), end: K this.max()): MapEntry<K, V>[] {
```

After writing this, I think we should make the end of the range be inclusive so that by default this will include the max key.  Same goes for values and keys.

Reply:
Ah yes I had the same initial approach, but unfortunately it results in an unexpected behavior where calling entries() will throw instead of returning an empty array when there are no keys in the tree. This happens because this.min() (and this.max()) throws if the tree is empty.